### ISSUE TITLE
Class Name Reporting - Strip off leading class names in VM

### ISSUE THREAD

**Issue by kenzieschmoll** [dart-lang/sdk#36999]
Class names in stack traces can get long and ugly when a class has several mixins. Using this name in the DevTools UI leads to problematically long strings.

Example: `_WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding._handlePersistentFrameCallback`

The important information here would be `_WidgetsFlutterBinding._handlePersistentFrameCallback`

Can we strip mixins off of the reported name? 

**EDIT**
**TL;DR: We should report the class name where the method is defined.**
For this example, we should report `C.method` instead of `A.method`.
```
main() {
  A().method();
}
class A extends B with C { }
mixin C {
  void method() {
    throw '!';
  }
}
```

**Comment by bkonyi** [comment]
I don't think this is something we'd want to do within the VM as it's not core functionality that needs to be provided by the VM itself. I'd also like to avoid introducing unnecessary c-string manipulation logic into the service which could lead to memory leaks or segfaults if we're not careful.

This processing is something that would probably be more easily implemented in Dart on the client side and could maybe be added to `package:vmservice` if it's something that multiple clients would find useful.

**Comment by jacob314** [comment]
To clarify the request is that the VM should stop generating long confusing stack frame names like
`_WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding._handlePersistentFrameCallback`
in the first place which would be more consistent with ddc and dart2js.
I believe
`RendererBinding._handlePersistentFrameCallback` which is the actual method being called is the correct name to display as that is consistent with how calls are displayed in other contexts.

@munificent @lrhn does the the Dart language specification indicate how these call frames should be displayed? DDC, Dart2JS and the VM are currently inconsistent on this case so it would be nice to standardize. If the specification currently recommend the `&` syntax I can file a bug to modify the spec to display simpler and more readable names for this case.

**Comment by kenzieschmoll** [comment]
Verified with an example that VM reports `Class&SuperClass&Mixin.method` where as dart2js reports `Class.method`

**Comment by bkonyi** [comment]
After some experimentation, `_WidgetsFlutterBinding._handlePersistentFrameCallback` actually isn't the correct name to report as `_handlePersistentFrameCallback` is actually defined in `RendererBinding`, so the name reported should technically be `RendererBinding._handlePersistentFrameCallback`.

What's happening here is the VM applies mixins individually, creating new classes for each individual mixin application. We then apply the next mixin to the new intermediate class until we've applied all the mixins. As a result, when we have a stack trace generated for a method `foo` defined in a mixin `M` and we have `class A with M1, M2, M, M3`, the frame will be reported as `_A&Object&M1&M2&M.foo` as `foo` is defined internally in the VM in intermediate class `_A&Object&M1&M2&M`.

Given that, the correct thing to do is to strip everything _before_ the last `&` as that's the class where the method is defined. For clarity, @rmacnak-google mentioned that other languages would report something like `A(M).foo` which generalizes to `$receiverType($typeOfMethodOwner).$method`, which would be more useful in my opinion.

**Comment by kenzieschmoll** [comment]
I agree that `A(M).foo` would be more useful than `M.foo`. If mixin `M` is used in several classes, `M.foo` misleadingly lumps all the call sites together.

**Comment by munificent** [comment]
> If mixin M is used in several classes, `M.foo` misleadingly lumps all the call sites together.

I think what users care about is where the method is declared, and not necessarily what receiver it happens to have been called on. The analogous case is inheritance. If I have:

```dart
class A {
  a() => throw "!";
}

class B extends A {}

main() {
  B().a();
}
```

It prints:

```
Unhandled exception:
!
#0      A.a (temp.dart:2:10)
#1      main (temp.dart:9:7)
#2      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:300:19)
#3      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:171:12)
```

Even though the receiver of the method was a B, it's only A that appears in the callstack, and that's fine. Users just want to be able to find the code where the error occurred.

I think this logic works equally well for mixins. Even with mixin application (the `class Foo = A with B` syntax), there's still always an explicit named entity for each thing that actually contains methods. You'll never need to show a mixin application in a stack trace since it doesn't actually define any new members. So you should always be able to just use the name of the mixin declaration.

**Comment by lrhn** [comment]
I agree with @munificent here. Showing *only* the mixin name in the stack trace is both reasonable and consistent.

We don't print the run-time type of the receiver of other methods, just the declaration point of the method. We could do that:
```
#0      A.a (temp.dart:2:10) on B (temp.dart:5:0)
```
It would probably help debugging. It would certainly slow down stack-trace generation too.

We should also not *prohibit* tools from using the mixin application class "name" (mixin application classes are anonymous except when declared with `class Name = ...;`) if that is the information they have readily available. One implementation strategy for mixins is to copy the method onto the mixin application class, and if the compiler did that, then the run-time may need to do extra work in order to find the original mixin declaration name for a stack trace.

The language specification does not specify anything about how stack traces look. We can't, we use JavaScript stack traces when compiled to JavaScript, and we do not have any control over how they look. All we can do is to encourage the tools to do something that is useful to people who need to debug the source of the stack trace.

**Comment by kenzieschmoll** [comment]
Filed an [issue](https://github.com/dart-lang/sdk/issues/37387) on dart2js to fix this there.

**Comment by kenzieschmoll** [comment]
@bkonyi can we implement this change in the VM to strip off the leading class names from reporting?

**Comment by bkonyi** [comment]
Will do! I'll hopefully get to it in the next week or so.
